### PR TITLE
Added syncpointspec.json tests to RTDB

### PR DESCRIPTION
**Pull request by maneesht**

### CODE REVIEW

File: packages/database/test/helpers/syncpoint-util.ts [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+      );
+    }
+    return false;
+  }
+  hasAnyCallback(): boolean {
+    return true;
+  }
+}
+
+function patchFakeAuthFunctions(app: FirebaseApp) {
```
Comment (maneesht) [inline packages/database/test/helpers/syncpoint-util.ts:175]: See if this is needed

File: packages/database/test/helpers/syncpoint-util.ts [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+        }
+      }
+    };
+    let currentSpec;
+    console.log('Running ' + testSpec.name);
+    let currentWriteId = 0;
+    const registrations = {};
+
+    for (let i = 0; i < testSpec.steps.length; ++i) {
+      const spec = unsafeClone(testSpec.steps[i]);
```
Comment (maneesht) [inline packages/database/test/helpers/syncpoint-util.ts:457]: Add a TODO to not clone this and create a separate object that deals with its own concerns

File: packages/database/test/syncpoint.test.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { SyncPointTestParser } from './helpers/syncpoint-util';
+import syncPointSpecs from './helpers/syncPointSpec.json';
```
Comment (maneesht) [inline packages/database/test/syncpoint.test.ts:19]: Eventually, we should split this up and support multiple specs

File: packages/database/test/helpers/syncpoint-util.ts [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+          i++;
+        }
+
+        // Step 2: foreach in actual, asserting location
+        for (let j = 0; j < i; ++j) {
+          const actualEventData = currentActual[j];
+          const eventFn =
+            actualEventData.eventRegistration.getEventRunner(actualEventData);
+          eventFn();
+          console.log(currentSpec);
```
Comment (maneesht) [inline packages/database/test/helpers/syncpoint-util.ts:328]: Delete this

File: packages/database/test/helpers/syncpoint-util.ts [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+}
+
+export class SyncPointTestParser {
+  app: FirebaseApp;
+  listens_: any = {};
+  listenProvider_: ListenProvider;
+  private syncTree_: SyncTree;
+  constructor() {
+    this.app = initializeApp(
+      { databaseURL: 'http://tests.fblocal.com:9000' },
```
Comment (IanWyszynski) [inline packages/database/test/helpers/syncpoint-util.ts:237]: It looks like these run against an emulator. For the future, it'd be nice if we could point them at prod/staging as well.

**Review by IanWyszynski — APPROVED** [review]
LGTM after nits discussed in-person are addressed.

**Pull request merged**